### PR TITLE
refactor: make DataSourceType an Enum

### LINKED ISSUE #6609: Make DataSourceType an enum

### PR DESCRIPTION
**Pull request by domoritz**
Use an enum with string values, instead of using an union type
with string literals. This should give better compile time type checking
and error messages.

Fixes #6609

Replaces https://github.com/vega/vega-lite/pull/6684/

### CODE REVIEW

File: examples/compiled/area_density_stacked.vg.json [hunk 1]
```diff
@@ -44,7 +44,7 @@
       "type": "group",
       "from": {
         "facet": {
-          "name": "faceted_path_main",
+          "name": "faceted_path_Main",
```
Comment (kanitw) [inline examples/compiled/area_density_stacked.vg.json:47]: Better keep the lowercase

Comment (znation) [reply examples/compiled/area_density_stacked.vg.json:47]: Yup, I am realizing that is the cause of most of my test failures. Will keep lowercase.

**Comment by domoritz** [comment]
Screen Shot 2020-06-25 at 17 24 08

**Comment by znation** [comment]
@domoritz I see code coverage checks are failing, but I did add a unit test for all of the new code added (as far as I can tell). Not sure why it's reporting that coverage is still too low. Otherwise, I think this PR is ready to go in. Should we override the coverage requirement here, or is there any other test coverage I should add?

**Comment by domoritz** [comment]
Since the code that isn't covered isn't new, we can accept your patch. I'll take a look if you think it's ready.

**Comment by znation** [comment]
@domoritz Yes, I think it's ready. Please take a look, thanks!

**Comment by domoritz** [comment]
Thank you

**Pull request merged**